stellar-contract-env-host/src/events.rs

```diff
@@ -1,39 +1,227 @@
-#![allow(dead_code)]
+use log::debug;
+use stellar_contract_env_common::{
+    xdr::{self},
+    RawVal, Status,
+};
 
-use arrayvec::ArrayVec;
-use stellar_contract_env_common::RawVal;
+use tinyvec::TinyVec;
+
+#[cfg(feature = "vm")]
+use crate::HostError;
+#[cfg(feature = "vm")]
+use stellar_contract_env_common::xdr::{ScUnknownErrorCode, ScVmErrorCode};
 
 // TODO: update this when ContractEvent shows up in the XDR defns.
 // TODO: optimize storage on this to use pools / bumpalo / etc.
 #[derive(Clone, Debug)]
-pub(crate) enum HostEvent {
+pub enum HostEvent {
+    #[allow(dead_code)]
     Contract(/*ContractEvent*/),
-    // Same size as LargeDebug but doesn't chew up a heap allocation.
-    SmallDebug {
-        msg: &'static str,
-        // ArrayVec<RawVal, 2> is 3 words: RawVal, RawVal, len.
-        args: ArrayVec<RawVal, 2>,
-    },
-    // Arbitrary set of k/v arguments.
-    LargeDebug {
-        msg: &'static str,
-        // Vec is 3 words: ptr, len, capacity.
-        args: Vec<RawVal>,
-    },
+    Debug(DebugEvent),
 }
 
 #[derive(Clone, Debug, Default)]
 pub struct Events(Vec<HostEvent>);
 
 impl Events {
-    pub(crate) fn add_debug_event(&mut self, msg: &'static str, in_args: &[RawVal]) {
-        let debug = if in_args.len() <= 2 {
-            let args: ArrayVec<RawVal, 2> = ArrayVec::from_iter(in_args.iter().cloned());
-            HostEvent::SmallDebug { msg, args }
-        } else {
-            let args = in_args.into();
-            HostEvent::LargeDebug { msg, args }
+    // Records the smallest variant of a debug HostEvent it can, returning the size of the
+    // in_args slice (for charging to a budget).
+    pub fn record_debug_event(&mut self, de: DebugEvent) -> u64 {
+        let len = de.args.len();
+        self.0.push(HostEvent::Debug(de));
+        len as u64
+    }
+
+    pub fn dump_to_debug_log(&self) {
+        for e in self.0.iter() {
+            match e {
+                HostEvent::Contract() => debug!("Contract event: <TBD>"),
+                HostEvent::Debug(e) => debug!("Debug event: {:?}", e),
+            }
+        }
+    }
+}
+
+/// A cheap record type to store in the events buffer for diagnostic reporting
+/// when something goes wrong. Should cost very little even when enabled. See
+/// [host::debug_event] for normal use.
+#[derive(Clone, Debug)]
+pub struct DebugEvent {
+    pub msg: &'static str,
+    pub args: TinyVec<[RawVal; 2]>,
+}
+
+impl DebugEvent {
+    pub fn new() -> Self {
+        Self {
+            msg: "",
+            args: Default::default(),
+        }
+    }
+
+    pub fn msg(mut self, msg: &'static str) -> Self {
+        self.msg = msg;
+        self
+    }
+
+    pub fn arg<T>(mut self, arg: T) -> Self
+    where
+        RawVal: From<T>,
+    {
+        self.args.push(arg.into());
```

Comment:
This call may panic if overflow happens, should we prevent that from happening by doing the condition check?

Reply:
TinyVec is not ArrayVec -- it automatically overflows to the heap; the panic here would only happen if we literally run out of memory, which is a ubiquitous error we don't generally guard against.